File: packs/dsa-linked-lists-py/challenges/06-remove-duplicates-sorted.test.py
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


def to_list(head):
    result = []
    while head:
        result.append(head.val)
        head = head.next
    return result


def from_list(lst):
    dummy = ListNode(0)
    curr = dummy
    for val in lst:
        curr.next = ListNode(val)
        curr = curr.next
    return dummy.next


from solution import delete_duplicates


def test_remove_duplicates_basic():
    head = from_list([1, 1, 2])
    result = delete_duplicates(head)
    assert to_list(result) == [1, 2]


def test_remove_duplicates_multiple():
    head = from_list([1, 1, 2, 3, 3])
    result = delete_duplicates(head)
    assert to_list(result) == [1, 2, 3]


def test_remove_duplicates_empty():
    result = delete_duplicates(None)
    assert result is None


def test_remove_duplicates_no_duplicates():
    head = from_list([1, 2, 3])
    result = delete_duplicates(head)
    assert to_list(result) == [1, 2, 3]


def test_remove_duplicates_all_same():
    head = from_list([1, 1, 1, 1])
    result = delete_duplicates(head)
    assert to_list(result) == [1]


def test_remove_duplicates_single():
    head = from_list([1])
    result = delete_duplicates(head)
    assert to_list(result) == [1]

File: packs/dsa-linked-lists-ts/challenges/06-remove-duplicates-sorted.test.ts
import { join } from 'node:path';
import { writeChallengeToTmp, importModule } from '../../test-helpers.js';

class ListNode {
  val: number;
  next: ListNode | null;
  constructor(val = 0, next: ListNode | null = null) {
    this.val = val;
    this.next = next;
  }
}

function toList(head: ListNode | null): number[] {
  const result: number[] = [];
  while (head) {
    result.push(head.val);
    head = head.next;
  }
  return result;
}

function fromList(arr: number[]): ListNode | null {
  const dummy = new ListNode(0);
  let curr = dummy;
  for (const val of arr) {
    curr.next = new ListNode(val);
    curr = curr.next;
  }
  return dummy.next;
}

const CHALLENGE = join(import.meta.dirname, '06-remove-duplicates-sorted.json');

describe('Remove Duplicates from Sorted List', () => {
  let deleteDuplicates: (head: ListNode | null) => ListNode | null;
  let cleanup: () => void;

  beforeAll(async () => {
    const tmp = writeChallengeToTmp(CHALLENGE);
    cleanup = tmp.cleanup;
    const mod = await importModule<{ deleteDuplicates: typeof deleteDuplicates }>(
      join(tmp.tmpDir, 'solution.ts'),
    );
    deleteDuplicates = mod.deleteDuplicates;
  });

  afterAll(() => cleanup());

  it('removes duplicates from a basic list', () => {
    expect(toList(deleteDuplicates(fromList([1, 1, 2])))).toEqual([1, 2]);
  });

  it('removes multiple groups of duplicates', () => {
    expect(toList(deleteDuplicates(fromList([1, 1, 2, 3, 3])))).toEqual([1, 2, 3]);
  });

  it('returns null for an empty list', () => {
    expect(deleteDuplicates(null)).toBeNull();
  });

  it('returns the same list when there are no duplicates', () => {
    expect(toList(deleteDuplicates(fromList([1, 2, 3])))).toEqual([1, 2, 3]);
  });

  it('reduces all-same list to a single node', () => {
    expect(toList(deleteDuplicates(fromList([1, 1, 1, 1])))).toEqual([1]);
  });

  it('returns a single-element list unchanged', () => {
    expect(toList(deleteDuplicates(fromList([1])))).toEqual([1]);
  });
});
